test/core/test/fixtures/timers.suite.ts

```diff
@@ -338,7 +338,7 @@ describe('FakeTimers', () => {
     })
   })
 
-  describe('runAllTimersAsync', () => {
+  describe.skipIf('TODO Enable this')('runAllTimersAsync', () => {
```

Comment:
> For me it's almost always something related to timer mocks

To talk about timers... 😅 
Every other test is now passing (even on CI) but this one. This is throwing the `Error: [birpc] timeout on calling "onTaskUpdate"` error. I think that error used to come up on `main` branch's CI maybe a month ago but now it's gone from there.

I can reproduce this error locally on MacOS when using [`stress(1)` CLI tool](https://github.com/resurrecting-open-source-projects/stress) to simulate CPU load. That tool is available on `brew` for easy installation.

```
# In one terminal
$ stress --cpu 8 --io 4 --vm 2 --vm-bytes 128M --timeout 200s

# In second terminal. Run couple of times if it doesn't come up immediately.
$ cd vitest/test/core 
$ pnpm run test --run --no-threads --allow-only
```

I've been looking at Vitest's `onTaskUpdate` usage for long now and cannot understand why that error is coming up. Runner should be `await`'ing that call to finish. But I guess when timers are mocked, all kinds of unexpected behavior can come up. The test case is even passing `global.Promise` to `fake-timers`. 🤔 

I think I'll take a look into `@sinonjs/fake-timers` sources next.

Reply:
Looks like this is actually a bug on `main` and doesn't exactly relate to this PR's changes.

When a test case mocks `process.nextTick` during test case, any execution that relies on `nextTick` will not be run. When `useRealTimers()` is called, those pending tasks are still not executed. In `birpc` we are storing the original `setTimeout`, `clearTimeout` and `Math.random` to prevent similar cases. I'm not sure if we even can keep internal dependencies running properly when test execution shares the global context with runner. VM isolation might help here?

Here is test case that will fail on latest Vitest release or `main` branch when run as `vitest --no-threads`. It might need to be ran multiple times, or slightly increase the timeout value. 

```ts
import { test, vi } from "vitest";

test("leaky process mocking", async () => {
  const ogSetTimeout = setTimeout;

  // Fake `process.nextTick` and execute longer than 10ms. Use heavy looping or simple `setTimeout`
  vi.useFakeTimers({ toFake: ["nextTick"] });

  process.stdout.write("\nThis is executed\n\n");
  await new Promise((resolve) => ogSetTimeout(resolve, 100)); // onTaskUpdate uses 10ms
  process.stdout.write("\nThis is also executed\n\n");

  vi.useRealTimers();

  process.stdout.write("\nTest finished\n\n");
});
```

The test case runs for 60s and `birpc` timeouts. Also it fails to report failed test case from child process now. 

<img width="372" alt="image" src="https://github.com/vitest-dev/vitest/assets/14806298/9d363d82-4648-475b-8baf-7f83f2613957">



Here is minimal-ish reproduction using just `@sinonjs/fake-timers` and `node:child_process`. When main process uses `process.send`, and subprocess has `process.nextTick` mocked, the `process.on("message", callback)` is not executed. And after `useRealTimers()` call those pending tasks are not executed at all.

<details>

```js
/*
 * Sends multiple messages from main process to subprocess.
 * Subprocess will fail if it does not receive all the messages.
 * Between the message sending, subprocess will install fake-timers.
 * After a while, it will uninstall fake-timers.
 *
 * Requires:
 * pnpm i @sinonjs/fake-timers@11.0.0
 */

import { fork } from "node:child_process";
import { fileURLToPath } from "node:url";

const filename = fileURLToPath(import.meta.url);
const isMainThread = !Boolean(process.env.CHILD_PROCESS);

const INTERVAL = 1000; // Time in ms between messages
const LOOPS = Array(10).fill(0).map((_, i) => i); // Amount of messages to be sent

if (isMainThread) {
  console.log("[main] running");
  const subprocess = fork(filename, { env: { CHILD_PROCESS: "true" } });
  const onExit = new Promise((r) => subprocess.on("exit", r));
  const onReady = new Promise((r) => subprocess.on("message", (m) => m === "ready" && r()));

  await onReady;
  console.log("[main] subprocess ready\n");

  // Send amount of LOOPS to subprocess with INTERVAL delay in between
  for (const index of LOOPS) {
    await new Promise((r) => setTimeout(r, INTERVAL));
    console.log("[main] sending message", index);
    subprocess.send({ loop: index });
  }

  // Let process exit when it's done
  subprocess.send({ exit: "graceful" });

  await onExit;
  console.log("[main] subprocess exited");
} else {
  console.log("[subprocess] running");

  const ogSetTimeout = setTimeout;
  const ogClearTimeout = clearTimeout;

  // These are used to indicate whether all messages were not received
  const timeouts = LOOPS.map((i) =>
    ogSetTimeout(() => { throw new Error(`Timeout ${i}`) }, INTERVAL * LOOPS.length * 1.5)
  );

  async function onMessage(message) {
    // Remove timeout from given LOOP message. If any messages are missed, a timeout error is expected
    if (message?.loop !== undefined) {
      console.log("[subprocess] received loop", message.loop);
      ogClearTimeout(timeouts[message.loop]);
      return;
    }

    if (message?.exit === "graceful") {
      process.off("message", onMessage);
      return console.log("[subprocess] exitting gracefully");
    }

    console.log("[subprocess] received unknown message", message);
  }

  process.on("message", onMessage);

  // Setup fake-timers
  const { withGlobal } = await import("@sinonjs/fake-timers");

  class FakeDate extends Date {}
  const global = { Date: FakeDate, clearTimeout, process, setTimeout, Promise };
  const fakeTimers = withGlobal(global);

  // Install fake-timers after 3s. After another 3s uninstall them.
  ogSetTimeout(() => {
    console.log("[subprocess] installing fake-timers");
    const clock = fakeTimers.install({
      now: Date.now(),
      toFake: Object.keys(fakeTimers.timers),
    });

    setTimeout(() => {
      console.log("[subprocess] uninstalling fake-timers");
      clock.uninstall();
    }, 3000);
  }, 3000);

  process.send("ready");
}
```

</details>

Does this analysis sound correct? The amount of debugging across `vitest`, `birpc`, `tinypool` and `@sinonjs/fake-timers` was a lot. 😫

Reply:
Sounds correct, I also had a lot of problems with `nextTick` the first time child process wad implemented. I fixed it by restoring original `nextTick` before calling rpc method, does it not work anymore?

Reply:
> fixed it by restoring original `nextTick` before calling rpc method

Good point, I was unaware of this. This is done by `withSafeTimers`, right?

That one is not `await`'ing the given callback. Calls of `birpc` are Promises when they are not specified in its `eventNames` config. By adding `await` here the failing minimal test case above is constantly passing:

https://github.com/vitest-dev/vitest/blob/91853bb204ed8dfc3750d54b1b79409168ea440b/packages/vitest/src/runtime/rpc.ts#L30-L31

However that one still doesn't fix all issues, as now `rpc().snapshotSaved` seems to timeout on Windows. I'll continue debugging. 🙃

Reply:
The main reason why it wasn’t awaiting is because the function itself was running synchronously and returned a promise, so all globals should’ve been called by that time. Maybe implementation changed in the meantime? Or maybe it was never working as intended 👀

Reply:
> function itself was running synchronously and returned a promise, so all globals should’ve been called by that time.

When the rpc method is called synchronously, it means that only the subprocess's `process.send` is ensured to be run using original `nextTick`. The response coming from main process is not received in `process.on('message', callback)` if `nextTick` is mocked. We'll need to `await` the whole rpc call to make sure response is received from main process. 

But this also means that any code in user's test case that runs between this cycle is running without mocked `nextTick`.

Reply:
I would just disallow mocking `nextTick` to be honest

Reply:
I'm starting to think that would be the best option too. This is driving me crazy. Debugging NodeJS when such low level internals are mocked is not fun. 
We could raise an error if `useFakeTimers({toFake: ['nextTick']})` is called when `node:child_process` is used. I don't think that many users are even interested in `nextTick` mocking.